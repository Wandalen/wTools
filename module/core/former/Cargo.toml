[package]
name = "former"
version = "2.7.0"
edition = "2021"
authors = [
  "Kostiantyn Wandalen <wandalen@obox.systems>",
  "Dmytro Kryvoruchko <dmytro.kr@obox.systems>",
]
license = "MIT"
readme = "Readme.md"
documentation = "https://docs.rs/former"
repository = "https://github.com/Wandalen/wTools/tree/master/module/core/former"
homepage = "https://github.com/Wandalen/wTools/tree/master/module/core/former"
description = """
A flexible implementation of the Builder pattern supporting nested builders and collection-specific subformers.
"""
categories = [ "algorithms", "development-tools" ]
keywords = [ "fundamental", "general-purpose", "builder-pattern" ]

[lints]
workspace = true

[package.metadata.docs.rs]
features = [ "full" ]
all-features = false

[features]

no_std = [ "former_types/no_std", "collection_tools/no_std" ]
use_alloc = [ "no_std", "former_types/use_alloc", "collection_tools/use_alloc" ]

# no_std = [ "collection_tools/no_std" ]
# use_alloc = [ "no_std", "collection_tools/use_alloc" ]

default = [
  "enabled",
  "derive_former",
  "derive_components",
  "derive_component_from",
  "derive_component_assign",
  "derive_components_assign",
  "derive_from_components",
  "types_former",
  "types_component_assign",
]
full = [
  "enabled",
  "derive_former",
  "derive_components",
  "derive_component_from",
  "derive_component_assign",
  "derive_components_assign",
  "derive_from_components",
  "types_former",
  "types_component_assign",
]
enabled = [ "former_meta/enabled", "former_types/enabled" ]

derive_former = [ "former_meta/derive_former", "types_former" ]
derive_components = [ "former_meta/derive_components", "derive_component_assign", "derive_components_assign", "derive_component_from", "derive_from_components" ]
derive_component_assign = [ "former_meta/derive_component_assign", "types_component_assign" ]
derive_components_assign = [ "derive_component_assign", "former_meta/derive_components_assign" ]
derive_component_from = [ "former_meta/derive_component_from" ]
derive_from_components = [ "former_meta/derive_from_components" ]

types_former = [ "former_types/types_former" ]
types_component_assign = [ "former_types/types_component_assign" ]

[dependencies]
former_meta = { workspace = true }
former_types = { workspace = true }
# collection_tools = { workspace = true, features = [ "collection_constructors" ] }


[dev-dependencies]
test_tools = { workspace = true, features = [ "full" ] }
collection_tools = { workspace = true, features = [ "collection_constructors" ] }
