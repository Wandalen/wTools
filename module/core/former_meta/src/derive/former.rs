
use super::*;
use iter_tools::{ Itertools, process_results };
use macro_tools::{ attr, diag, generics, container_kind, typ, Result };
use proc_macro2::TokenStream;

///
/// Descriptor of a field.
///

#[ allow( dead_code ) ]
struct FormerField< 'a >
{
  pub attrs : Attributes,
  pub vis : &'a syn::Visibility,
  pub ident : &'a syn::Ident,
  pub colon_token : &'a Option< syn::token::Colon >,
  pub ty : &'a syn::Type,
  pub non_optional_ty : &'a syn::Type,
  pub is_optional : bool,
  pub of_type : container_kind::ContainerKind,
}

///
/// Attributes of the field.
///

struct Attributes
{
  default : Option< AttributeDefault >,
  setter : Option< AttributeSetter >,
  // #[ allow( dead_code ) ]
  subformer : Option< AttributeFormer >,
  alias : Option< AttributeAlias >,
}

impl Attributes
{
  fn parse( attributes : & Vec< syn::Attribute > ) -> Result< Self >
  {
    let mut default = None;
    let mut setter = None;
    let mut subformer = None;
    let mut alias = None;
    for attr in attributes
    {
      let key_ident = attr.path().get_ident()
      .ok_or_else( || syn_err!( attr, "Expects an attribute of format #[ attribute( val ) ], but got:\n  {}", qt!{ #attr } ) )?;
      let key_str = format!( "{}", key_ident );
      match key_str.as_ref()
      {
        "default" =>
        {
          match attr.meta
          {
            syn::Meta::List( ref meta_list ) =>
            {
              default.replace( syn::parse2::< AttributeDefault >( meta_list.tokens.clone() )? );
            },
            _ => return_syn_err!( attr, "Expects an attribute of format #[ attribute( val ) ], but got:\n  {}", qt!{ #attr } ),
          }
        }
        "setter" =>
        {
          match attr.meta
          {
            syn::Meta::List( ref meta_list ) =>
            {
              setter.replace( syn::parse2::< AttributeSetter >( meta_list.tokens.clone() )? );
            },
            _ => return_syn_err!( attr, "Expects an attribute of format #[ attribute( val ) ], but got:\n  {}", qt!{ #attr } ),
          }
          // let attr_setter = syn::parse2::< AttributeSetter >( attr.tokens.clone() )?;
          // setter.replace( attr_setter );
        }
        "subformer" =>
        {
          match attr.meta
          {
            syn::Meta::List( ref meta_list ) =>
            {
              subformer.replace( syn::parse2::< AttributeFormer >( meta_list.tokens.clone() )? );
            },
            _ => return_syn_err!( attr, "Expects an attribute of format #[ attribute( val ) ], but got:\n  {}", qt!{ #attr } ),
          }
          // let attr_former = syn::parse2::< AttributeFormer >( attr.tokens.clone() )?;
          // subformer.replace( attr_former );
        }
        "alias" =>
        {
          match attr.meta
          {
            syn::Meta::List( ref meta_list ) =>
            {
              alias.replace( syn::parse2::< AttributeAlias >( meta_list.tokens.clone() )? );
            },
            _ => return_syn_err!( attr, "Expects an attribute of format #[ attribute( val ) ], but got:\n  {}", qt!{ #attr } ),
          }
          // let attr_alias = syn::parse2::< AttributeAlias >( attr.tokens.clone() )?;
          // alias.replace( attr_alias );
        }
        "doc" =>
        {
        }
        _ =>
        {
          return Err( syn_err!( attr, "Unknown attribute {}", qt!{ #attr } ) );
        }
      }
    }

    Ok( Attributes { default, setter, subformer, alias } )
  }
}

///
/// Attribute to hold information about method to call after form.
///
/// `#[ perform( fn after1< 'a >() -> Option< &'a str > ) ]`
///

#[ allow( dead_code ) ]
struct AttributeFormAfter
{
  // paren_token : syn::token::Paren,
  signature : syn::Signature,
}

impl syn::parse::Parse for AttributeFormAfter
{
  fn parse( input : syn::parse::ParseStream< '_ > ) -> Result< Self >
  {
    // let input2;
    Ok( Self
    {
      // paren_token : syn::parenthesized!( input2 in input ),
      // signature : input2.parse()?,
      signature : input.parse()?,
    })
  }
}

///
/// Attribute to hold information about default value.
///
/// `#[ default( 13 ) ]`
///

#[ allow( dead_code ) ]
struct AttributeDefault
{
  // eq_token : syn::Token!{ = },
  // paren_token : syn::token::Paren,
  expr : syn::Expr,
}

impl syn::parse::Parse for AttributeDefault
{
  fn parse( input : syn::parse::ParseStream< '_ > ) -> Result< Self >
  {
    // let input2;
    Ok( Self
    {
      // paren_token : syn::parenthesized!( input2 in input ),
      // eq_token : input.parse()?,
      // expr : input2.parse()?,
      expr : input.parse()?,
    })
  }
}

///
/// Attribute to enable/disable setter generation.
///
/// `#[ setter( false ) ]`
///

#[ allow( dead_code ) ]
struct AttributeSetter
{
  // paren_token : syn::token::Paren,
  condition : syn::LitBool,
}

impl syn::parse::Parse for AttributeSetter
{
  fn parse( input : syn::parse::ParseStream< '_ > ) -> Result< Self >
  {
    // let input2;
    Ok( Self
    {
      // paren_token : syn::parenthesized!( input2 in input ),
      // condition : input2.parse()?,
      condition : input.parse()?,
    })
  }
}

///
/// Attribute to enable/disable former generation.
/// Also known as subformers, used for aggregation relationship, when a struct holds another struct, which needs to be build by invoking multiple methods
/// Typical example is a struct holding a `Vec`
///
/// `#[ subformer( former::VectorSubformer ) ]`
///

#[ allow( dead_code ) ]
struct AttributeFormer
{
  // paren_token : syn::token::Paren,
  expr : syn::Type,
}

impl syn::parse::Parse for AttributeFormer
{
  fn parse( input : syn::parse::ParseStream< '_ > ) -> Result< Self >
  {
    // let input2;
    Ok( Self
    {
      // paren_token : syn::parenthesized!( input2 in input ),
      // expr : input2.parse()?,
      expr : input.parse()?,
    })
  }
}

///
/// Attribute to create alias.
///
/// `#[ alias( name ) ]`
///

#[ allow( dead_code ) ]
struct AttributeAlias
{
  // paren_token : syn::token::Paren,
  alias : syn::Ident,
}

impl syn::parse::Parse for AttributeAlias
{
  fn parse( input : syn::parse::ParseStream< '_ > ) -> Result< Self >
  {
    // let input2;
    Ok( Self
    {
      // paren_token : syn::parenthesized!( input2 in input ),
      // alias : input2.parse()?,
      alias : input.parse()?,
    })
  }
}

///
/// Is type under Option.
///

fn is_optional( ty : &syn::Type ) -> bool
{
  typ::type_rightmost( ty ) == Some( "Option".to_string() )
}

///
/// Extract the first parameter of the type if such exist.
///

fn parameter_internal_first( ty : &syn::Type ) -> Result< &syn::Type >
{
  typ::type_parameters( ty, 0 ..= 0 )
  .first()
  .copied()
  .ok_or_else( || syn_err!( ty, "Expects at least one parameter here:\n  {}", qt!{ #ty } ) )
}

///
/// Generate fields for initializer of a struct setting each field to `None`.
///
/// Used for initializing a Container, where on initialization all fields are None. User can alter them through builder pattern
///
/// ### Basic use-case. of output
///
/// ```ignore
/// int_1 : core::option::Option::None,
/// string_1 : core::option::Option::None,
/// int_optional_1 : core::option::Option::None,
/// ```
///

#[ inline( always ) ]
fn field_none_map( field : &FormerField< '_ > ) -> TokenStream
{
  let ident = Some( field.ident.clone() );
  let tokens = qt! { ::core::option::Option::None };
  let ty2 : syn::Type = syn::parse2( tokens ).unwrap();

  qt!
  {
    #ident : #ty2
  }
}

///
/// Generate field of the former for a field of the structure
///
/// Used to generate a Container
///
/// ### Basic use-case. of output
///
/// ```ignore
/// pub int_1 : core::option::Option< i32 >,
/// pub string_1 : core::option::Option< String >,
/// pub int_optional_1 :  core::option::Option< i32 >,
/// pub string_optional_1 : core::option::Option< String >,
/// ```
///

#[ inline( always ) ]
fn field_optional_map( field : &FormerField< '_ > ) -> TokenStream
{
  let ident = Some( field.ident.clone() );
  let ty = field.ty.clone();

  // let ty2 = if is_optional( &ty )
  let ty2 = if field.is_optional
  {
    qt! { #ty }
  }
  else
  {
    qt! { ::core::option::Option< #ty > }
  };

  qt!
  {
    pub #ident : #ty2
  }

}

///
/// Generate code converting a field of the former to the field of the structure.
///
/// In simple terms, used on `form()` call to unwrap contained values from the former's storage.
/// Will try to use default values if no values supplied by the former and the type implements `Default` trait.
///
/// ### Generated code will look similar to this :
///
/// ```ignore
/// let int_1 : i32 = if self.storage.int_1.is_some()
/// {
///   // if int_1 is optional
///   Some( self.storage.int_1.take().unwrap() )
///
///   // if int_1 isn't optional
///   self.storage.int_1.take().unwrap()
/// }
/// else
/// {
///   // if int_1 is optional and has default
///   Some( i32::default().into() )
///
///   // if int_1 is optional and doesn't have default
///   None
///
///   // if int_1 isn't optional and has default
///   i32::default().into()
///
///   // if int_1 isn't optional and hasn't default
///   panic!( "Field 'int_1' isn't initialized" )
/// };
/// ```
///

#[ inline( always ) ]
fn field_form_map( field : &FormerField< '_ > ) -> Result< TokenStream >
{
  let ident = field.ident;
  let ty = field.ty;
  let default = field.attrs.default.as_ref()
  .map( | attr_default | &attr_default.expr );

  let tokens = if field.is_optional
  {

    let _else = match default
    {
      None =>
      {
        qt!
        {
          ::core::option::Option::None
        }
      }

      Some( default_val ) =>
      {
        qt!
        {
          ::core::option::Option::Some( ( #default_val ).into() )
        }
      }
    };

    qt!
    {
      let #ident = if self.#ident.is_some()
      {
        ::core::option::Option::Some( self.#ident.take().unwrap() )
      }
      else
      {
        #_else
      };
    }

  }
  else
  {

    let _else = match default
    {
      None =>
      {
        let panic_msg = format!( "Field '{}' isn't initialized", ident );
        qt!
        {
          {
            // By hardly utilizing deref coercion, we achieve conditional trait implementation
            trait MaybeDefault< T >
            {
              fn maybe_default( self : &Self ) -> T { panic!( #panic_msg ) }
            }

            // Panic on non-`Default` types
            impl< T > MaybeDefault< T >
            for &::core::marker::PhantomData< T >
            {}

            // Return default value on `Default`` types
            impl< T > MaybeDefault< T >
            for ::core::marker::PhantomData< T >
            where T : ::core::default::Default,
            {
              fn maybe_default( self : &Self ) -> T
              {
                T::default()
              }
            }

            // default if `impl Default`, otherwise - panic
            ( &::core::marker::PhantomData::< #ty > ).maybe_default()
          }
        }
      }
      Some( default_val ) =>
      {
        qt!
        {
          ( #default_val ).into()
        }
      }
    };

    qt!
    {
      let #ident = if self.#ident.is_some()
      {
        self.#ident.take().unwrap()
      }
      else
      {
        #_else
      };
    }

  };

  Ok( tokens )
}

///
/// Extract name of a field out.
///

#[ inline( always ) ]
fn field_name_map( field : &FormerField< '_ > ) -> syn::Ident
{
  field.ident.clone()
}

///
/// Generate a former setter for the field.
///
/// If aliases provided, also generate aliases
///
/// # Example of output
/// ```ignore
/// #[ doc = "Setter for the 'int_1' field." ]
/// #[ inline ]
/// pub fn int_1< Src >( mut self, src : Src ) -> Self
/// where
///   Src : ::core::convert::Into< i32 >,
/// {
///   debug_assert!( self.int_1.is_none() );
///   self.storage.int_1 = ::core::option::Option::Some( src.into() );
///   self
/// }
///
/// /// #[ doc = "Setter for the 'int_1' field." ]
/// #[ inline ]
/// pub fn int_1_alias< Src >( mut self, src : Src ) -> Self
/// where
///   Src : ::core::convert::Into< i32 >,
/// {
///   debug_assert!( self.int_1.is_none() );
///   self.storage.int_1 = ::core::option::Option::Some( src.into() );
///   self
/// }
/// ```

#[ inline ]
fn field_setter_map( field : &FormerField< '_ > ) -> Result< TokenStream >
{
  let ident = &field.ident;

  if let Some( setter_attr ) = &field.attrs.setter
  {
    if !setter_attr.condition.value()
    {
      return Ok( qt!{ } );
    }
  }

  let non_optional_ty = &field.non_optional_ty;
  // Either subformer or ordinary setter.
  let setter_tokens = if let Some( subformer_ty ) = &field.attrs.subformer
  {
    subformer_field_setter( ident, ident, non_optional_ty, &subformer_ty.expr )
    // field_setter( ident, ident, non_optional_ty )
  }
  else
  {
    field_setter( ident, ident, non_optional_ty )
  };

  let r = if let Some( alias_attr ) = &field.attrs.alias
  {
    let alias_tokens = field_setter( ident, &alias_attr.alias, non_optional_ty );
    let token = qt!
    {
      #setter_tokens
      #alias_tokens
    };
    Ok( token )
  }
  else
  {
    Ok( setter_tokens )
  };

  // tree_print!( r.as_ref().unwrap() );
  r
}

///
/// Generate a single setter for the 'field_ident' with the 'setter_name' name.
///
/// Used as a helper function for field_setter_map(), which generates all alias setters
///
/// # Example of output
/// ```ignore
/// #[ doc = "Setter for the 'int_1' field." ]
/// #[ inline ]
/// pub fn int_1< Src >( mut self, src : Src ) -> Self
/// where
///   Src : ::core::convert::Into< i32 >,
/// {
///   debug_assert!( self.int_1.is_none() );
///   self.storage.int_1 = ::core::option::Option::Some( src.into() );
///   self
/// }
/// ```

#[ inline ]
fn field_setter
(
  field_ident : &syn::Ident,
  setter_name : &syn::Ident,
  non_optional_type : &syn::Type,
)
-> TokenStream
{
  let doc = format!
  (
    "Setter for the '{}' field.",
    field_ident,
  );

  qt!
  {
    #[ doc = #doc ]
    #[ inline ]
    pub fn #setter_name< Src >( mut self, src : Src ) -> Self
    where Src : ::core::convert::Into< #non_optional_type >,
    {
      debug_assert!( self.storage.#field_ident.is_none() );
      self.storage.#field_ident = ::core::option::Option::Some( src.into() );
      self
    }
  }
}

///
/// Generate a sub-former setter for the 'field_ident' with the 'setter_name' name.
///
/// # Example of generated code
///
/// ```ignore
/// pub fn hashmap_strings_1( mut self ) -> former::HashMapSubformer
/// <
///   String,
///   String,
///   std::collections::HashMap< String, String >,
///   Struct1Former,
///   impl Fn( std::collections::HashMap< String, String >, core::option::Option< Self > ) -> Self
/// >
/// {
///   let formed = self.hashmap_strings_1.take();
///   let on_end = | formed : std::collections::HashMap< String, String >, mut former : core::option::Option< Self > | -> Self
///   {
///     former.hashmap_strings_1 = Some( formed );
///     former
///   };
///   former::HashMapSubformer::begin( formed, self, on_end )
/// }
/// ```

#[ inline ]
fn subformer_field_setter
(
  field_ident : &syn::Ident,
  setter_name : &syn::Ident,
  non_optional_type : &syn::Type,
  subformer_type : &syn::Type,
)
-> TokenStream
{
  let doc = format!
  (
    "Subformer setter for the '{}' field.",
    field_ident
  );

  // tree_print!( non_optional_type );
  // code_print!( non_optional_type );
  let params = typ::type_parameters( &non_optional_type, .. );
  // params.iter().for_each( | e | println!( "{}", qt!( #e ) ) );

  qt!
  {
    #[ doc = #doc ]
    #[ inline ]
    pub fn #setter_name( mut self ) -> #subformer_type
    <
      #( #params, )*
      #non_optional_type,
      Self,
      impl Fn( #non_optional_type, core::option::Option< Self > ) -> Self,
    >
    {
      let formed = self.storage.#setter_name.take();
      let on_end = | formed : #non_optional_type, former : core::option::Option< Self > | -> Self
      {
        let mut former = former.unwrap();
        former.storage.#setter_name = Some( formed );
        former
      };
      #subformer_type::begin( formed, Some( self ), on_end )
    }
  }

}

///
/// Generate documentation for the former.
///

fn doc_generate( name_ident : &syn::Ident ) -> ( String, String )
{

  let doc_former_mod = format!
  (
r#" Implementation of former for [{}].
"#,
    name_ident
  );

  let doc_example1 =
r#"
use former::Former;
#[ derive( Former ) ]
pub struct Struct1
{
  #[default( 31 ) ]
  field1 : i32,
}
"#;

  let doc_former_struct = format!
  (
r#" Object to form [{}]. If field's values is not set then default value of the field is set.

For specifying custom default value use attribute `default`. For example:
```
{}
```
"#,
    name_ident, doc_example1
  );

  ( doc_former_mod, doc_former_struct )
}

//

///
/// Generate parts, used for generating `perform()`` method.
///
/// Similar to `form()`, but will also invoke function from `perform` attribute, if specified.
///
/// # Example of returned tokens :
///
/// ## perform :
/// return result;
///
/// ## perform_output :
/// < T : ::core::default::Default >
///
/// ## perform_generics :
/// Vec< T >

pub fn performer< 'a >
(
  name_ident : &syn::Ident,
  generics_ty : &syn::TypeGenerics< '_ >,
  attrs : impl Iterator< Item = &'a syn::Attribute >,
)
-> Result< ( TokenStream, TokenStream, TokenStream ) >
{

  let mut perform = qt!
  {
    return result;
  };
  let mut perform_output = qt!{ #name_ident #generics_ty };
  let mut perform_generics = qt!{};
  for attr in attrs
  {
    if let Some( ident ) = attr.path().get_ident()
    {
      let ident_string = format!( "{}", ident );
      if ident_string == "perform"
      {
        match attr.meta
        {
          syn::Meta::List( ref meta_list ) =>
          {
            // default.replace( syn::parse2::< AttributeDefault >( meta_list.tokens.clone() )? );
            // let attr_perform = syn::parse2::< AttributeFormAfter >( attr.tokens.clone() )?;
            let attr_perform = syn::parse2::< AttributeFormAfter >( meta_list.tokens.clone() )?;
            let signature = &attr_perform.signature;
            let generics = &signature.generics;
            perform_generics = qt!{ #generics };
            let perform_ident = &signature.ident;
            let output = &signature.output;
            if let syn::ReturnType::Type( _, boxed_type ) = output
            {
              perform_output = qt!{ #boxed_type };
            }
            perform = qt!
            {
              return result.#perform_ident();
            };
          },
          _ => return_syn_err!( attr, "Expects an attribute of format #[ attribute( val ) ], but got:\n  {}", qt!{ #attr } ),
        }
      }
    }
    else
    {
      return_syn_err!( "Unknown structure attribute:\n{}", qt!{ attr } );
    }
  }

  Ok( ( perform, perform_output, perform_generics ) )
}

//

///
/// Generate the whole Former ecosystem
///
/// Output examples can be found in [docs to former crate](https://docs.rs/former/latest/former/)
///

pub fn former( input : proc_macro::TokenStream ) -> Result< TokenStream >
{

  let original_input = input.clone();
  let ast = match syn::parse::< syn::DeriveInput >( input )
  {
    Ok( syntax_tree ) => syntax_tree,
    Err( err ) => return Err( err ),
  };
  let has_debug = attr::has_debug( ast.attrs.iter() )?;
  let example_of_custom_setter = false;

  /* names */

  let name_ident = &ast.ident;
  let former_name = format!( "{}Former", name_ident );
  let former_name_ident = syn::Ident::new( &former_name, name_ident.span() );
  let former_storage_name = format!( "{}FormerStorage", name_ident );
  let former_storage_name_ident = syn::Ident::new( &former_storage_name, name_ident.span() );

  /* generic parameters */

  let generics = &ast.generics;
  let ( generics_impl, generics_ty, generics_where ) = generics.split_for_impl();
  let _generics_params = generics::params_names( generics ).params;
  let generics_params = if _generics_params.len() == 0
  {
    qt!{}
  }
  else
  {
    qt!{ #_generics_params, }
  };

// __Formed

  // add embedded generic parameters
  let mut extra_generics : syn::Generics = parse_quote!
  {
    < __FormerContext = (), __FormerEnd = former::ReturnStorage >
    // < __FormerContext = #name_ident #generics_ty, __FormerEnd = former::ReturnStorage >
  };
  extra_generics.where_clause = parse_quote!
  {
    where __FormerEnd : former::FormingEnd< #former_storage_name_ident, __FormerContext, #name_ident #generics_ty >,
  };
  // xxx : write helper to fix bug with where
  let generics_of_former = generics::merge( &generics, &extra_generics );
  let ( generics_of_former_impl, generics_of_former_ty, generics_of_former_where ) = generics_of_former.split_for_impl();
  let generics_of_former_with_defaults = generics_of_former.params.clone();
  // macro_tools::code_print!( generics_of_former_with_defaults );
  // macro_tools::code_print!( extra_generics );

  /* structure attribute */

  let ( perform, perform_output, perform_generics ) = performer
  (
    &name_ident,
    &generics_ty,
    ast.attrs.iter(),
  )?;

  /* */

  let fields = match ast.data
  {
    syn::Data::Struct( ref data_struct ) => match data_struct.fields
    {
      syn::Fields::Named( ref fields_named ) =>
      {
        &fields_named.named
      },
      _ => return Err( syn_err!( ast, "Unknown format of data, expected syn::Fields::Named( ref fields_named )\n  {}", qt!{ #ast } ) ),
    },
    _ => return Err( syn_err!( ast, "Unknown format of data, expected syn::Data::Struct( ref data_struct )\n  {}", qt!{ #ast } ) ),
  };

  let former_fields : Vec< Result< FormerField< '_ > > > = fields.iter().map( | field |
  {
    let attrs = Attributes::parse( &field.attrs )?;
    let vis = &field.vis;
    let ident = field.ident.as_ref()
    .ok_or_else( || syn_err!( field, "Expected that each field has key, but some does not:\n  {}", qt!{ #field } ) )?;
    let colon_token = &field.colon_token;
    let ty = &field.ty;
    let is_optional = is_optional( ty );
    let of_type = container_kind::of_optional( ty ).0;
    let non_optional_ty : &syn::Type = if is_optional { parameter_internal_first( ty )? } else { ty };
    let former_field = FormerField { attrs, vis, ident, colon_token, ty, non_optional_ty, is_optional, of_type };
    Ok( former_field )
  }).collect();

  let former_fields : Vec< _ > = process_results( former_fields, | iter | iter.collect() )?;

  let ( fields_none, fields_optional, fields_form, fields_names, fields_setter )
  : ( Vec< _ >, Vec< _ >, Vec< _ >, Vec< _ >, Vec< _ > )
  = former_fields.iter().map( | former_field |
  {(
    field_none_map( former_field ),
    field_optional_map( former_field ),
    field_form_map( former_field ),
    field_name_map( former_field ),
    field_setter_map( former_field ),
  )}).multiunzip();

  let ( _doc_former_mod, doc_former_struct ) = doc_generate( name_ident );
  let fields_setter : Vec< _ > = process_results( fields_setter, | iter | iter.collect() )?;
  let fields_form : Vec< _ > = process_results( fields_form, | iter | iter.collect() )?;

  let result = qt!
  {

    #[ automatically_derived ]
    impl #generics_impl #name_ident #generics_ty
    #generics_where
    {
      ///
      /// Make former, variation of builder pattern to form structure defining values of fields step by step.
      ///
      #[ inline( always ) ]
      pub fn former() -> #former_name_ident < #generics_params #name_ident #generics_ty, former::ReturnStorage >
      {
        #former_name_ident :: < #generics_params #name_ident #generics_ty, former::ReturnStorage > :: new()
      }
    }

    // xxx : rename to storage
    #[ doc = "Container of a corresponding former." ]
    pub struct #former_storage_name_ident #generics_ty
    #generics_where
    {
      #(
        /// A field
        #fields_optional,
      )*
    }

    #[ automatically_derived ]
    impl #generics_impl #former_storage_name_ident #generics_ty
    #generics_where
    {

      ///
      /// Finish setting options and return formed entity.
      ///
      /// `perform` has no effect on method `form`, but change behavior and returned type of method `perform`.
      ///
      #[ inline( always ) ]
      pub fn preform( mut self ) -> #name_ident #generics_ty
      {
        #( #fields_form )*
        let result = #name_ident
        {
          #( #fields_names, )*
        };
        return result;
      }

    }

    impl #generics_impl ::core::default::Default for #former_storage_name_ident #generics_ty
    #generics_where
    {

      #[ inline( always ) ]
      fn default() -> Self
      {
        Self
        {
          #( #fields_none, )*
        }
      }

    }

    #[ doc = #doc_former_struct ]
    #[ automatically_derived ]
    pub struct #former_name_ident < #generics_of_former_with_defaults >
    #generics_of_former_where
    {
      storage : #former_storage_name_ident #generics_ty,
      context : core::option::Option< __FormerContext >,
      on_end : core::option::Option< __FormerEnd >,
    }

    #[ automatically_derived ]
    impl #generics_of_former_impl #former_name_ident #generics_of_former_ty
    #generics_of_former_where
    {

      ///
      /// Finish setting options and return formed entity.
      ///
      /// `perform` has no effect on method `form`, but change behavior and returned type of method `perform`.
      ///
      #[ inline( always ) ]
      pub fn preform( self ) -> #name_ident #generics_ty
      {
        self.storage.preform()
        // #( #fields_form )*
        // let result = #name_ident
        // {
        //   #( #fields_names, )*
        // };
        // return result;
      }

      ///
      /// Finish setting options and call perform on formed entity.
      ///
      /// If `perform` defined then associated method is called and its result returned instead of entity.
      /// For example `perform()` of structure with : `#[ perform( fn after1() -> &str > )` returns `&str`.
      ///
      #[ inline( always ) ]
      pub fn perform #perform_generics ( self ) -> #perform_output
      {
        let result = self.form();
        #perform
      }

      ///
      /// Begin the process of forming. Expects context of forming to return it after forming.
      ///
      #[ inline( always ) ]
      pub fn begin
      (
        mut storage : core::option::Option< #former_storage_name_ident #generics_ty >,
        context : core::option::Option< __FormerContext >,
        on_end : __FormerEnd,
      ) -> Self
      {
        if storage.is_none()
        {
          storage = Some( ::core::default::Default::default() );
        }
        Self
        {
          storage : storage.unwrap(),
          context : context,
          on_end : ::core::option::Option::Some( on_end ),
        }
      }

      ///
      /// End the process of forming returning original context of forming.
      ///
      #[ inline( always ) ]
      pub fn form( mut self ) -> #name_ident #generics_ty
      {
        self.end()
      }

      // xxx : improve documentation
      ///
      /// End the process of forming returning original context of forming.
      ///
      #[ inline( always ) ]
      pub fn end( mut self ) -> #name_ident #generics_ty
      {
        let on_end = self.on_end.take().unwrap();
        let context = self.context.take();
        let storage = self.form();
        on_end.call( storage, context )
      }

      #(
        #fields_setter
      )*

    }

    #[ automatically_derived ]
    // impl #generics_impl #former_name_ident < #generics_params #name_ident #generics_ty, former::ReturnStorage >
    impl #generics_impl #former_name_ident < #generics_params (), former::ReturnStorage >
    #generics_where
    {

      ///
      /// Construct new instance of former with default parameters.
      ///
      #[ inline( always ) ]
      pub fn new() -> Self
      {
        // #former_name_ident :: < #generics_params #name_ident #generics_ty, former::ReturnStorage > :: begin
        Self :: begin
        (
          None,
          None,
          former::ReturnStorage,
        )
      }

    }

  };

  if has_debug
  {
    diag::debug_report_print( "derive : Former", original_input, &result );
  }

  // xxx : implement
  if example_of_custom_setter
  {
    let _example =
r#"
impl< Context, End > UserProfileFormer< Context, End >
where
  End : former::FormingEnd< UserProfileFormerStorage, Context, UserProfile >,
{
  pub fn age< Src >( mut self, src : Src ) -> Self
  where
    Src : Into< i32 >,
  {
    debug_assert!( self.age.is_none() );
    self.storage.age = ::core::option::Option::Some( src.into() );
    self
  }
}
"#;
  }

  Ok( result )
}

// xxx : explain concept of Storage
