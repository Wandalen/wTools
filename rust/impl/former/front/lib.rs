#![ doc( html_logo_url = "https://raw.githubusercontent.com/Wandalen/wTools/master/asset/img/logo_v3_trans_square.png" ) ]
#![ warn( rust_2018_idioms ) ]
#![ warn( missing_debug_implementations ) ]
#![ warn( missing_docs ) ]

// #![ feature( type_name_of_val ) ]
// #![ feature( trace_macros ) ]

//!
//! Former - variation of builder pattern.
//!

#![ doc = include_str!( concat!( env!( "CARGO_MANIFEST_DIR" ), "/Readme.md" ) ) ]

/// Namespace with dependencies.
pub mod dependency
{
  // #[ cfg( any( feature = "runtime", feature = "former_runtime" ) ) ]
  pub use former_runtime;
  // #[ cfg( any( feature = "meta", feature = "former_meta" ) ) ]
  pub use former_meta;
}

/// Protected namespace of the module.
pub mod protected
{
  pub use super::orphan::*;
}

pub use protected::*;

/// Parented namespace of the module.
pub mod orphan
{
  pub use super::exposed::*;
}

/// Exposed namespace of the module.
pub mod exposed
{
  pub use super::prelude::*;
  // #[ cfg( any( feature = "runtime", feature = "former_runtime" ) ) ]
  pub use former_runtime as runtime;
  // #[ cfg( any( feature = "meta", feature = "former_meta" ) ) ]
  pub use former_meta as derive;
  // #[ cfg( any( feature = "meta", feature = "former_meta" ) ) ]
  pub use derive::*;
}

/// Prelude to use essentials: `use my_module::prelude::*`.
pub mod prelude
{
}
