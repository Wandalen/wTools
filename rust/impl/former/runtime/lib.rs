#![ doc( html_logo_url = "https://raw.githubusercontent.com/Wandalen/wTools/master/asset/img/logo_v3_trans_square.png" ) ]
#![ warn( rust_2018_idioms ) ]
#![ warn( missing_debug_implementations ) ]
#![ warn( missing_docs ) ]

//!
//! Former - variation of builder pattern. Implementation of its runtime.
//!

#![ doc = include_str!( concat!( env!( "CARGO_MANIFEST_DIR" ), "/Readme.md" ) ) ]

/// Namespace with dependencies.
pub mod dependency
{
}

/// Former of a fector.
mod vector;
/// Former of a hash map.
mod hash_map;
/// Former of a hash set.
mod hash_set;

/// Protected namespace of the module.
pub mod protected
{
  pub use super::exposed::*;
  pub use super::vector::*;
  pub use super::hash_map::*;
  pub use super::hash_set::*;
}

pub use protected::*;

/// Exposed namespace of the module.
pub mod exposed
{
  pub use super::prelude::*;
}

/// Prelude to use essentials: `use my_module::prelude::*`.
pub mod prelude
{
}
