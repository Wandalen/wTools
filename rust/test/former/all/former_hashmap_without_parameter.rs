use former::Former;

struct HashMap< T >
{
  f1 : T,
}

#[derive( Former )]
pub struct Struct1
{
  pub string_slice_1 : HashMap< i32 >,
}

fn main()
{
}
