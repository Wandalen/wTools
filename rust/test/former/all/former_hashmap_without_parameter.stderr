error: Expects at least two parameters here:
         HashMap < i32 >
  --> ../../../rust/test/former/all/former_hashmap_without_parameter.rs:11:24
   |
11 |   pub string_slice_1 : HashMap< i32 >,
   |                        ^^^^^^^^^^^^^^
